Tests/UnitTests/PaymentsCoreTests/NetworkingTests/APIClient_Tests.swift

```diff
@@ -0,0 +1,189 @@
+
+        apiClient.fetch(endpoint: mockRequestResponse) { result, _ in
+            guard case .failure(.decodingError) = result else {
+                XCTFail("Expect NetworkingError.decodingError for invalid response")
+                return
+            }
+
+            expect.fulfill()
+        }
+        waitForExpectations(timeout: 1)
+    }
+
+    func testFetch_withEmptyResponse_vendsSuccessfully() {
+        let expect = expectation(description: "Get empty response type for mock request")
+        let emptyRequest = MockEmptyRequestResponse()
+
+        URLProtocolMock.requestResponses.append(emptyRequest)
+
+        apiClient.fetch(endpoint: emptyRequest) { result, _ in
+            guard case .success = result else {
+                XCTFail("Expected successful empty response")
+                return
+            }
+            expect.fulfill()
+        }
+        waitForExpectations(timeout: 1)
+    }
+
+    func testFetch_withError_failsSuccessfully() {
+        let expect = expectation(description: "should vend error to client call site")
+        let serverError = NSError(
+            domain: URLError.errorDomain,
+            code: NSURLErrorBadServerResponse,
+            userInfo: nil
+        )
+
+        let mockRequest = MockRequestResponse(
+            request: AccessTokenRequest(clientID: "123"),
+            statusCode: 400,
+            error: serverError
+        )
+
+        URLProtocolMock.requestResponses.append(mockRequest)
+
+        apiClient.fetch(endpoint: mockRequest) { result, _ in
+            guard case .failure(.networkingError) = result else {
+                XCTFail("Should receive a networking error response")
+                return
+            }
+            expect.fulfill()
+        }
+        waitForExpectations(timeout: 1)
+    }
+
+    func testFetch_noResponse_vendsNoReponseError() {
+        let expect = expectation(description: "Should receive bad URLResponse error")
+        let noReponseRequest = MockNoReponseRequest()
+
+        URLProtocolMock.requestResponses.append(noReponseRequest)
+
+        apiClient.fetch(endpoint: noReponseRequest) { result, _ in
+            guard case .failure(.badURLResponse) = result else {
+                XCTFail("Expected bad URLResponse error")
+                return
+            }
+            expect.fulfill()
+        }
+        waitForExpectations(timeout: 1)
+    }
+
+    func testFetch_noURLRequest_vendsNoURLRequestError() {
+        let expect = expectation(description: "Should receive noURLRequest error")
+
+        // Mock request whose API object does not vend a URLRequest
+        let noURLRequest = MockNoURLRequest()
+
+        URLProtocolMock.requestResponses.append(noURLRequest)
+
+        apiClient.fetch(endpoint: noURLRequest) { result, _ in
+            guard case .failure(.noURLRequest) = result else {
+                XCTFail("Expected bad URLResponse error")
+                return
+            }
```

Comment:
These tests should assert that the error type is equal to  `.noURLRequest`. This gives the test less room to potentially give a false positive.

Reply:
to do positive assertion, the NetworkingError'd need to conform to `Equatable`.
That'd allow us to do `XCTEqual(theError, theExpectedErrorCase)` for enum without associated value.
But still, for cases with associated value like `connectionIssue(Error)` or `parsingError(Error)`, we'd still need to do this `guard case`.

IMO the guard case here already assert that the error here needs to be `.noURLRequest` otherwise it fails, so I don't think we'll ever give a false positive.
What do you think we leave this as is?

Reply:
The `XCTAssertEqual` should work fine on the Enum that conforms to Error. I think where it goes wonky is because we have these two cases `case networkingError(Error)` & `case decodingError(Error)` which take an arg. If these cases didn't take params, then we could do `XCTAssertEqual(error, .noURLRequest)` etc.

I think we should be able to assert directly on our error types. Our merchants tests & apps should be able to do this as well, without needed to do this workaround via the guard.

Also, I know we have a TODO for follow up later to re-implement our Error type. We could punt on this until then and include a note on it in the TODO comment?

Reply:
Without associated value, It still needs to be `Equatable`

![Screen Shot 2021-09-27 at 2 46 50 PM](https://user-images.githubusercontent.com/80476442/134967240-79a6a9ee-0856-45ec-99c2-1f3230df3cbc.png)

But yes, I'll put a TODO comment there for this.